File: AudiobooksTV/Core/RomanNumerals.swift
import Foundation

enum RomanNumerals {
    private static let values: [(Int, String)] = [
        (1000, "M"), (900, "CM"), (500, "D"), (400, "CD"),
        (100, "C"), (90, "XC"), (50, "L"), (40, "XL"),
        (10, "X"), (9, "IX"), (5, "V"), (4, "IV"), (1, "I"),
    ]

    /// Parses a Roman numeral (case-insensitive). Returns nil if the string
    /// contains non-numeral characters or is not consumable greedily.
    static func parse(_ s: String) -> Int? {
        let s = s.uppercased()
        guard !s.isEmpty, s.allSatisfy({ "MDCLXVI".contains($0) }) else { return nil }
        var total = 0
        var index = s.startIndex
        for (value, symbol) in values {
            while s[index...].hasPrefix(symbol) {
                total += value
                index = s.index(index, offsetBy: symbol.count)
                if index == s.endIndex { return total }
            }
        }
        return index == s.endIndex ? total : nil
    }
}

File: web/src/lib/romanNumerals.ts
const VALUES: Array<[number, string]> = [
  [1000, "M"], [900, "CM"], [500, "D"], [400, "CD"],
  [100, "C"], [90, "XC"], [50, "L"], [40, "XL"],
  [10, "X"], [9, "IX"], [5, "V"], [4, "IV"], [1, "I"],
];

/**
 * Parses a Roman numeral (case-insensitive). Returns null if the string
 * contains non-numeral characters or is not consumable greedily.
 */
export function parseRoman(s: string): number | null {
  const upper = s.toUpperCase();
  if (upper.length === 0 || ![...upper].every((c) => "MDCLXVI".includes(c))) return null;
  let total = 0;
  let index = 0;
  for (const [value, symbol] of VALUES) {
    while (upper.startsWith(symbol, index)) {
      total += value;
      index += symbol.length;
      if (index === upper.length) return total;
    }
  }
  return index === upper.length ? total : null;
}
